lxd/device/gpu_sriov.go

```diff
@@ -0,0 +1,297 @@
+	if err != nil {
+		return vfPCIDev, err
+	}
+
+	// Unbind VF device from the host so that the settings will take effect when we rebind it.
+	err = pci.DeviceUnbind(vfPCIDev)
+	if err != nil {
+		return vfPCIDev, err
+	}
+
+	revert.Add(func() { pci.DeviceProbe(vfPCIDev) })
+
+	// Register VF device with vfio-pci driver so it can be passed to VM.
+	err = pci.DeviceDriverOverride(vfPCIDev, "vfio-pci")
+	if err != nil {
+		return vfPCIDev, err
+	}
+
+	// Record original driver used by VF device for restore.
+	volatile["last_state.pci.driver"] = vfPCIDev.Driver
+
+	revert.Success()
+
+	return vfPCIDev, nil
+}
+
+// getVFDevicePCISlot returns the PCI slot name for a PCI virtual function device.
+func (d *gpuSRIOV) getVFDevicePCISlot(parentPCIAddress string, vfID string) (pci.Device, error) {
+	ueventFile := fmt.Sprintf("/sys/bus/pci/%s/device/virtfn%s/uevent", parentPCIAddress, vfID)
+	pciDev, err := pci.ParseUeventFile(ueventFile)
+	if err != nil {
+		return pciDev, err
+	}
+
+	return pciDev, nil
+}
+
+func (d *gpuSRIOV) findFreeVirtualFunction(parentDev pci.Device) (int, error) {
+	// Get number of currently enabled VFs.
+	sriovNumVFs := fmt.Sprintf("/sys/bus/pci/devices/%s/sriov_numvfs", parentDev.SlotName)
+
+	sriovNumVfsBuf, err := ioutil.ReadFile(sriovNumVFs)
+	if err != nil {
+		return 0, err
+	}
+	sriovNumVfsStr := strings.TrimSpace(string(sriovNumVfsBuf))
+	sriovNum, err := strconv.Atoi(sriovNumVfsStr)
+	if err != nil {
+		return 0, err
+	}
+
+	vfID := -1
+
+	for i := 0; i < sriovNum; i++ {
+		pciDev, err := pci.ParseUeventFile(fmt.Sprintf("/sys/bus/pci/devices/%s/virtfn%d", parentDev.SlotName, i))
+		if err != nil {
+			return 0, err
+		}
+
+		if pciDev.Driver == "" {
+			vfID = i
+			break
+		}
+	}
+
+	if vfID == -1 {
+		return 0, fmt.Errorf("All virtual functions on parent device %q seem to be in use", parentDev.ID)
+	}
+
+	return vfID, nil
+}
+
+// Stop is run when the device is removed from the instance.
+func (d *gpuSRIOV) Stop() (*deviceConfig.RunConfig, error) {
+	runConf := deviceConfig.RunConfig{
+		PostHooks: []func() error{d.postStop},
+	}
+
+	return &runConf, nil
+}
+
+// postStop is run after the device is removed from the instance.
+func (d *gpuSRIOV) postStop() error {
+	defer d.volatileSet(map[string]string{
+		"last_state.created":    "",
+		"last_state.vf.id":      "",
+		"last_state.pci.driver": "",
+	})
+
+	v := d.volatileGet()
+
+	err := d.restoreSriovParent(v)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// restoreSriovParent restores SR-IOV parent device settings when removed from an instance using the
+// volatile data that was stored when the device was first added with setupSriovParent().
+func (d *gpuSRIOV) restoreSriovParent(volatile map[string]string) error {
+	// Nothing to do if we don't know the original device name or the VF ID.
+	if volatile["last_state.vf.id"] == "" || (d.config["pci"] == "" && d.config["id"] == "" && d.config["vendorid"] == "" && d.config["productid"] == "") {
```

Comment:
@monstermunchkin did you also need to change `last_state.vf.parent` is defined here (assuming the key is needed at all based on my comment above)?